tools/src/main/java/org/apache/kafka/tools/TransactionalMessageCopier.java

```diff
@@ -387,7 +389,17 @@ public void onPartitionsAssigned(Collection<TopicPartition> partitions) {
                         long messagesSentWithinCurrentTxn = records.count();
 
                         ConsumerGroupMetadata groupMetadata = useGroupMetadata ? consumer.groupMetadata() : new ConsumerGroupMetadata(consumerGroup);
-                        producer.sendOffsetsToTransaction(consumerPositions(consumer), groupMetadata);
+                        try {
+                            producer.sendOffsetsToTransaction(consumerPositions(consumer), groupMetadata);
+                        } catch (KafkaException e) {
+                            // in case the producer gets stuck here, create a new one and continue the loop
+                            try { producer.close(Duration.ofSeconds(0)); } catch (Exception ignore) {}
+                            parsedArgs.getAttrs().put("transactionalId", parsedArgs.getString("transactionalId") + producerNumber++);
+                            producer = createProducer(parsedArgs);
+                            producer.initTransactions();
+                            resetToLastCommittedPositions(consumer);
```

Comment:
```suggestion
                            circutBreaker(); // in case the producer gets stuck here, create a new one and continue the loop
```

could give dedicated to this concern apply single responsibility principle, giving more focus to each own. Here its just about breaking the circut, how this is actually done seems to be some kind of (randomly) changing impl. detail.